equinox/nn/pool.py

```diff
@@ -392,3 +392,106 @@ def __call__(
+    if num_head_arrays != 0:
+        head_end_index = num_head_arrays * (channels // target_size + 1)
+        head_mean = jax.vmap(jnp.mean)(x[:head_end_index].reshape(num_head_arrays, -1))
+        tail_mean = jax.vmap(jnp.mean)(
+            x[head_end_index:].reshape(-1, channels // target_size)
+        )
+        mean = jnp.concatenate([head_mean, tail_mean])
+    else:
+        mean = jax.vmap(jnp.mean)(
+            jax.vmap(jnp.mean)(x.reshape(-1, channels // target_size))
+        )
+    return mean
+
+
+class AdaptiveAvgPool1D(Module):
+    """Adaptive 1D downsampling for a target shape."""
+
+    target_size: int = static_field()
+
+    def __init__(self, target_size: int):
+        """**Arguments:**
+
+        - `target_size`: The target output size.
+        """
+
+        self.target_size = target_size
+
+    def __call__(
+        self, x: Array, *, key: Optional["jax.random.PRNGKey"] = None
+    ) -> Array:
+        """**Arguments:**
+
+        - `x`: The input. Should be a JAX array of shape `(channels)`.
+        - `key`: Ignored; provided for compatibility with the rest of the Equinox API.
+            (Keyword only argument.)
+
+        **Returns:**
+
+        A JAX array of shape `(target_size)`.
+        """
+        mean = adaptive_avg_pool1d(x, self.target_size)
+        return mean
+
+
+class AdaptiveAvgPool2D(Module):
+    """Adaptive 2D downsampling for a target shape."""
+
+    target_size: Union[int, Sequence[int]] = static_field()
+
+    def __init__(self, target_size: Union[int, Sequence[int]]):
+        """**Arguments:**
+
+        - `target_size`: The target output size.
+        """
+        if isinstance(target_size, int):
+            target_size = (target_size, target_size)
+        self.target_size = target_size
+
+    def __call__(
+        self, x: Array, *, key: Optional["jax.random.PRNGKey"] = None
+    ) -> Array:
+        """**Arguments:**
+
+        - `x`: The input. Should be a JAX array of shape `(channel, dims)`.
+        - `key`: Ignored; provided for compatibility with the rest of the Equinox API.
+            (Keyword only argument.)
+
+        **Returns:**
+
+        A JAX array of shape `(target_size)`.
+        """
+        if x.ndim != 2:
+            raise ValueError(
+                f"2D input expected, received input with {x.ndim} dimensions."
+            )
+
+        channels, dims = x.shape
+        if channels < self.target_size[0] or dims < self.target_size[1]:
+            raise ValueError(
+                "`target_size` cannot be larger than the input channels."
+                f"Expected atleast {self.target_size} but received {channels}x{dims}."
+            )
+        target_channels, target_dims = self.target_size
+
+        x = jax.vmap(adaptive_avg_pool1d, in_axes=(0, None))(x, target_dims)
+        x = jax.vmap(adaptive_avg_pool1d, in_axes=(0, None))(
+            jnp.transpose(x), target_channels
+        )
+        return jnp.transpose(x)
```

Comment:
I think the transposing can be skipped here, in favour of just setting `in_axes=(1, None)`?

Reply:
Thanks for the suggestion! Still getting a hang of vmaps.